src/wp-includes/query.php

```diff
@@ -1117,15 +1105,40 @@ function _find_post_by_old_slug( $post_type ) {
 		$query .= $wpdb->prepare( ' AND DAYOFMONTH(post_date) = %d', get_query_var( 'day' ) );
 	}
 
+	$key          = md5( $query );
+	$cache_key = "_find_post_by_old_slug:$key";
```

Comment:
```suggestion
        $last_changed = wp_cache_get_last_changed( 'posts' );

	$key              = md5( $query );
	$cache_key = "_find_post_by_old_slug:$hash:$last_changed";
```

This needs to be like this to ensure that caches are invalidated.